pkg/certs/util.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package cert
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	"k8s.io/client-go/kubernetes"
+
+	"istio.io/istio/security/pkg/cmd"
+	k8ssecret "istio.io/istio/security/pkg/k8s/secret"
+	"istio.io/istio/security/pkg/pki/util"
+	"istio.io/pkg/log"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// CASecret defines plugged-in secret name
+	PlugInSecret = "cacerts"
+	// RootCertName defines plugged-in root cert file name
+	RootCertName = "root-cert.pem"
+	// CAKeyName defines plugged-in ca private key file name
+	CAKeyName = "ca-key.pem"
+	// CACertName  defines plugged-in ca cert file name
+	CACertName = "ca-cert.pem"
+	//CertChainName  defines plugged-in cert-chain file name
+	CertChainName = "cert-chain.pem"
+
+	caKeySize = 2048
+)
+
+// PlugInProvisionedCert plugin external CAs into Istio
+
+func PlugInProvisionedCert(client kubernetes.Interface, certDir, ns string, dryRun bool) error {
+	stat, err := os.Stat(certDir)
+	if err != nil {
+		return err
+	}
+	if stat.Mode().Perm() != 0777 {
+		return fmt.Errorf("specified certDir %q should have mode %#o, has %#o", certDir, os.ModePerm, stat.Mode())
+	}
+	rootCert, err := ioutil.ReadFile(filepath.Join(certDir, RootCertName))
+	if err != nil {
+		log.Warnf("can not read %q from %s", RootCertName, certDir)
+		return err
+	}
+	cakey, err := ioutil.ReadFile(filepath.Join(certDir, CAKeyName))
+	if err != nil {
+		log.Warnf("can not read %q from %s", CAKeyName, certDir)
+		return err
+	}
+	caCert, err := ioutil.ReadFile(filepath.Join(certDir, CACertName))
+	if err != nil {
+		log.Warnf("can not read %q from %s", CACertName, certDir)
+		return err
+	}
+	certChain, err := ioutil.ReadFile(filepath.Join(certDir, CertChainName))
+	if err != nil {
+		log.Warnf("can not read %q from %s", CertChainName, certDir)
+		return err
+	}
+
+	err = createSecret(client, ns, rootCert, cakey, caCert, certChain, dryRun)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//most parts are coming from security/tools/generate_cert/main.go
+// GenerateCertsAndKey is used to generate key and certs
+func GenerateCertsAndKey(client kubernetes.Interface, ns string, dryRun bool, outputDir, org string) error {
```

Comment:
I'm not clear about the benefits we get from this, and it's non-trivial. Maybe move this auto-generation function into another PR?